quant: add test util for testing off by one quantized tensors

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37781 quant: add test util for testing off by one quantized tensors**
* #37802 quant: clean up normalization channels_last handling

Summary:

Adds a test utility to verify two quantized tensors are close to each
other, but allowing for a percentage of off-by-scale errors. Useful
for quantized operators which have whitelisted numerics differences from
their floating point versions.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21385948](https://our.internmc.facebook.com/intern/diff/D21385948)

vkuzo committed:
quant: add test util for testing off by one quantized tensors

Summary:

Adds a test utility to verify two quantized tensors are close to each
other, but allowing for a percentage of off-by-scale errors. Useful
for quantized operators which have whitelisted numerics differences from
their floating point versions.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "quant: add test util for testing off by one quantized tensors"


Summary:

Adds a test utility to verify two quantized tensors are close to each
other, but allowing for a percentage of off-by-scale errors. Useful
for quantized operators which have whitelisted numerics differences from
their floating point versions.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21385948](https://our.internmc.facebook.com/intern/diff/D21385948)

[ghstack-poisoned]

vkuzo committed:
Update on "quant: add test util for testing off by one quantized tensors"


Summary:

Adds a test utility to verify two quantized tensors are close to each
other, but allowing for a percentage of off-by-scale errors. Useful
for quantized operators which have whitelisted numerics differences from
their floating point versions.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21385948](https://our.internmc.facebook.com/intern/diff/D21385948)

[ghstack-poisoned]

jerryzh168 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -968,6 +968,25 @@ def assertAlmostEqual(self, x, y, places=None, msg='', delta=None, allow_inf=Non
             prec = 10**(-places)
         self.assertEqual(x, y, msg, atol=prec, allow_inf=allow_inf)
 
+    def assertQuantizedClose(self, qx, qy, tol_off_by_y_scale=1e-2, message=''):
```
probably should be moved to common_quantization.py

vkuzo committed:
Update on "quant: add test util for testing off by one quantized tensors"


Summary:

Adds a test utility to verify two quantized tensors are close to each
other, but allowing for a percentage of off-by-scale errors. Useful
for quantized operators which have whitelisted numerics differences from
their floating point versions.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21385948](https://our.internmc.facebook.com/intern/diff/D21385948)

[ghstack-poisoned]

raghuramank100 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -970,6 +970,25 @@ def assertAlmostEqual(self, x, y, places=None, msg='', delta=None, allow_inf=Non
+
+        diff = torch.abs(qx.int_repr() - qy.int_repr())
+        off_by_one = torch.sum(diff == 1)
+        off_by_two_plus = torch.sum(diff > 1)
```
Is this off_by_one_plus?

vkuzo replied:
I meant `>=2`, but will reword

raghuramank100 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -970,6 +970,25 @@ def assertAlmostEqual(self, x, y, places=None, msg='', delta=None, allow_inf=Non
             prec = 10**(-places)
         self.assertEqual(x, y, msg, atol=prec, allow_inf=allow_inf)
 
+    def assertQuantizedClose(self, qx, qy, tol_off_by_y_scale=1e-2, message=''):
```
Also, can we call the argument fraction_of_one_off_error instead of tol_off_by_y_scale

vkuzo replied:
the current name is consistent with other similar arguments in this file, but i'm flexible, especially if this is moved elsewhere as Jerry is suggesting

vkuzo committed:
Update on "quant: add test util for testing off by one quantized tensors"


Summary:

Adds a test utility to verify two quantized tensors are close to each
other, but allowing for a percentage of off-by-scale errors. Useful
for quantized operators which have whitelisted numerics differences from
their floating point versions.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21385948](https://our.internmc.facebook.com/intern/diff/D21385948)

[ghstack-poisoned]

vkuzo committed:
Update on "quant: add test util for testing off by one quantized tensors"


Summary:

Adds a test utility to verify two quantized tensors are close to each
other, but allowing for a percentage of off-by-scale errors. Useful
for quantized operators which have whitelisted numerics differences from
their floating point versions.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21385948](https://our.internmc.facebook.com/intern/diff/D21385948)

[ghstack-poisoned]

vkuzo committed:
Update on "quant: add test util for testing off by one quantized tensors"


Summary:

Adds a test utility to verify two quantized tensors are close to each
other, but allowing for a percentage of off-by-scale errors. Useful
for quantized operators which have whitelisted numerics differences from
their floating point versions.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21385948](https://our.internmc.facebook.com/intern/diff/D21385948)

[ghstack-poisoned]
